salt/utils/etcd_util.py

```diff
@@ -497,11 +532,206 @@ def tree(self, path):
         return ret
 
 
+class EtcdApiV3Adapter(EtcdClient):
+    """
+    Since etcd3 has no concept of directories, this class leaves some methods unimplemented.
+
+    These are: ls and write_directory.
+    """
+
+    ENCODING = "UTF-8"
+
+    def __init__(self, opts, **kwargs):
+        if not HAS_ETCD_V3:
+            raise EtcdLibraryNotInstalled("Don't have etcd3-py, need to install it.")
+        log.debug("etcd_util has the libraries needed for etcd v3")
+
+        super().__init__(opts, **kwargs)
+
+        # etcd3-py uses verify instead of ca_cert
+        self.xargs["verify"] = self.xargs.pop("ca_cert", None)
+        self.client = etcd3.Client(host=self.host, port=self.port, **self.xargs)
+
+    def _decode_from_bytes(self, kv):
+        try:
+            kv.key = (
+                kv.key.decode(encoding=self.ENCODING)
+                if (kv.key and isinstance(kv.key, bytes))
+                else kv.key
+            )
+            kv.value = (
+                kv.value.decode(encoding=self.ENCODING)
+                if (kv.value and isinstance(kv.value, bytes))
+                else kv.value
+            )
+        except AttributeError as err:
+            log.debug("etcd3 decoding error: %s", err)
+        return kv
+
+    def watch(self, key, recurse=False, timeout=0, index=None):
+        ret = {"key": key, "value": None, "changed": False, "mIndex": 0, "dir": False}
+        result = self.read(
+            key, recursive=recurse, wait=True, timeout=timeout, waitIndex=index
+        )
+        if result is not None:
+            result = self._decode_from_bytes(result)
+            ret["key"] = result.key
+            ret["value"] = result.value
+            ret["mIndex"] = getattr(result, "mod_revision", 0)
+            ret["changed"] = True
+        else:
+            return {}
+        return ret
+
+    def get(self, key, recurse=False):
+        if not recurse:
+            result = self.read(key)
+            if isinstance(result, list):
+                return result.pop().value
+            return None
+        return self.tree(key)
+
+    def read(self, key, recursive=False, wait=False, timeout=None, waitIndex=None):
+        if not wait:
+            try:
+                result = self.client.range(key, prefix=recursive)
+                kvs = getattr(result, "kvs", None)
+                if kvs is None:
+                    log.error("etcd3 read: No values found for key %s", key)
+                else:
+                    for kv in kvs:
+                        kv = self._decode_from_bytes(kv)
```

Comment:
From what I saw here, https://github.com/Revolution1/etcd3-py/search?q=encoding, utf-8 is pretty much hardcoded in the v3 library.

Reply:
I see I wasn't terribly clear there. I guess what I mean to say is, consider the following:

```
with open('foo', 'r') as f:
    print(type(f.read()))
    
with open('foo', 'rb') as f:
    print(type(f.read()))
```

You'll get bytes from one and text/str from the other. Another example:

```
>>> msgpack.loads(msgpack.dumps({'foo': 'bar'}), raw=False)
{'foo': 'bar'}
>>> msgpack.loads(msgpack.dumps({'foo': 'bar'}), raw=True)
{b'foo': b'bar'}
```

That can be particularly important because in Python `b'foo' != 'foo'`

```
>>> msgpack.loads(msgpack.dumps({'foo': 'bar', b'foo': 'baz'}), raw=False)
{'foo': 'bar', b'foo': 'baz'}
>>> msgpack.loads(msgpack.dumps({'foo': 'bar', b'foo': 'baz'}), raw=True)
{b'foo': b'baz'}
```

Reply:
Oh. No.


Nonononononono.

Unfortunately, as we discovered on our call, etcd v3 API has an _incredible_ API, namely all keys are bytes -- period. Also values(?). So... that's uh... definitely a thing 😬

Reply:
Upon further discussion, I think the best approach here is to provide an optional (default true) encode key/values using msgpack. The result of msgpack encoding is a byte string, which will be correctly stored in etcd. The msgpack encoded bytestring can safely round-trip, and unlike the native etcd API, it *also* distinguishes between `b'foo'` and `'foo'`. Which is something that at least in Python definitely makes sense.

Of course, if someone is using etcd in an out-of-band sort of way (perhaps someone is doing `etcdctl put foo somevalue` for instance) then it won't make sense to use msgpack in this case. Instead we'll just have to let the ambiguity play out. But that will be an opt-in behavior -- and they might want to decode the key/value coming out, so they should be able to specify decoding and encoding behavior on a global scale as well as on a per-state or per-module call (which would override the global value, natch).